### PR TITLE
Gfx docs rework

### PR DESCRIPTION
**Pull request by icefoxen**
New PR to replace #1800 with something that's a little less messy, commit-wise.

There's still some things that need fixing, they are marked with `DOC TODO` comments.  But there's only a few of them, and they're all things I either don't understand or need more holistic knowledge to explain well.  I hope to revisit them and clean them up at some point but there's other things I want to do first.

I seem to have accidentally messed up the names of a couple functions, those will get fixed ASAP.

### CODE REVIEW

File: src/hal/src/command/graphics.rs [hunk 1]
```diff
@@ -144,39 +151,43 @@ pub enum BlitFilter {
...
 }
 
-///
+/// Parameters for an image blit operation, where a portion of one image
+/// is copied into another, possibly with scaling and filtering.
 #[derive(Clone, Debug)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct ImageBlit {
-    ///
+    /// Source image  and layers.
```
Comment (msiglreith) [inline src/hal/src/command/graphics.rs]: nit: whitespaces

File: src/hal/src/command/mod.rs [hunk 2]
```diff
@@ -17,24 +30,35 @@ pub use self::transfer::*;
...
+///
+/// Vulkan describes a primary command buffer as one which can be directly submitted
+/// to a queue, and can execute `Secondary` command buffers.
 pub enum Primary { }
 impl Level for Primary { }
+
 /// Indicates a secondary command buffer.
+///
+/// Vulkan describes a secondary command buffer as one which cannot be directly submitted
+/// to a queue, but can be executed by a primary command buffer.  This allows
```
Comment (msiglreith) [inline src/hal/src/command/mod.rs]: nit: whitespaces

File: src/hal/src/command/raw.rs [hunk 3]
```diff
@@ -18,74 +18,87 @@ use super::{
...
         const EMPTY = 0x0;
 
-        ///
+        /// Says that the command buffer will be recorded, submitted only once, and then reset and re-filled 
+        /// for another submission.
         const ONE_TIME_SUBMIT = 0x1;
 
-        ///
+        /// If set on a secondary command buffer, it says the command buffer takes place entirely inside
+        /// a render pass.  Ignored on primary command buffer.
```
Comment (msiglreith) [inline src/hal/src/command/raw.rs]: nit: whitespaces

File: src/hal/src/command/raw.rs [hunk 4]
```diff
@@ -95,23 +108,24 @@ pub trait RawCommandBuffer<B: Backend>: Clone + Any + Send + Sync {
...
     fn update_buffer(
         &mut self,
         buffer: &B::Buffer,
         offset: buffer::Offset,
         data: &[u8],
     );
 
-    /// Clear color image
+    /// Clears an image to the given color.
+    // Just calls `clear_color_raw` with some minor type conversion.
```
Comment (msiglreith) [inline src/hal/src/command/raw.rs]: missing `/` ?

File: src/hal/src/command/raw.rs [hunk 5]
```diff
@@ -160,15 +175,17 @@ pub trait RawCommandBuffer<B: Backend>: Clone + Any + Send + Sync {
...
     fn clear_attachments<T, U>(&mut self, clears: T, rects: U)
     where
         T: IntoIterator,
         T::Item: Borrow<AttachmentClear>,
         U: IntoIterator,
         U::Item: Borrow<Rect>;
 
-    ///
+    /// "Resolves" a multi-sampled image, converting it into a non-multi-sampled
+    /// image.  Takes an iterator of regions to apply the resolution to.
```
Comment (msiglreith) [inline src/hal/src/command/raw.rs]: nit: whitespaces

File: src/hal/src/command/raw.rs [hunk 6]
```diff
@@ -160,15 +175,17 @@ pub trait RawCommandBuffer<B: Backend>: Clone + Any + Send + Sync {
...
+    /// and clears the given rect's for *each* attachment.
     fn clear_attachments<T, U>(&mut self, clears: T, rects: U)
     where
         T: IntoIterator,
         T::Item: Borrow<AttachmentClear>,
         U: IntoIterator,
         U::Item: Borrow<Rect>;
 
-    ///
+    /// "Resolves" a multi-sampled image, converting it into a non-multi-sampled
```
Comment (msiglreith) [inline src/hal/src/command/raw.rs]: `non-multisampled`

File: src/hal/src/command/raw.rs [hunk 7]
```diff
@@ -243,14 +263,14 @@ pub trait RawCommandBuffer<B: Backend>: Clone + Any + Send + Sync {
         T: IntoIterator,
         T::Item: Borrow<Rect>;
 
-    ///
+    /// DOC TODO: I don't know how to explain this.
```
Comment (msiglreith) [inline src/hal/src/command/raw.rs]: Sets the stencil reference value for comparison operations and store operations.
Will be used on the LHS for stencil compare ops and as store value when the store op is `Reference`.

File: src/hal/src/command/raw.rs [hunk 8]
```diff
@@ -285,8 +305,13 @@ pub trait RawCommandBuffer<B: Backend>: Clone + Any + Send + Sync {
         )
     }
 
-    ///
-    fn begin_render_pass_raw<T>(
+    /// Begins recording commands for a render pass on the given framebuffer.
+    /// `render_area` is the section of the framebuffer to render,
+    /// `clear_values` is an iterator of `ClearValue`'s to use to use for
+    /// `clear_*` commands, one for each attachment of the render pass.
+    /// `first_subpass` specifies whether the rendering commands are provided
```
Comment (msiglreith) [inline src/hal/src/command/raw.rs]: first_subpass only indicates how commands are encoded for the first subpass.

File: src/hal/src/command/raw.rs [hunk 9]
```diff
@@ -345,7 +372,9 @@ pub trait RawCommandBuffer<B: Backend>: Clone + Any + Send + Sync {
         T: IntoIterator,
         T::Item: Borrow<B::DescriptorSet>;
 
-    /// Execute a workgroup in the compute pipeline.
+    /// Execute a workgroup in the compute pipeline.  `x`, `y` and `z` are the
```
Comment (msiglreith) [inline src/hal/src/command/raw.rs]: nit: whitespaces

File: src/hal/src/command/raw.rs [hunk 10]
```diff
@@ -374,7 +404,10 @@ pub trait RawCommandBuffer<B: Backend>: Clone + Any + Send + Sync {
         T: IntoIterator,
         T::Item: Borrow<BufferCopy>;
 
-    ///
+    /// Copies regions from the source to the destination images, which
+    /// have the given layouts.  No format conversion is done; the source and destination
+    /// `ImageLayout`'s **must** have the same sized texels (such as `Rgba8Unorm` and `R32`,
```
Comment (msiglreith) [inline src/hal/src/command/raw.rs]: `ImageLayout` doesn't affect texels, you probably mean image formats.

File: src/hal/src/command/raw.rs [hunk 11]
```diff
@@ -408,22 +441,38 @@ pub trait RawCommandBuffer<B: Backend>: Clone + Any + Send + Sync {
         T: IntoIterator,
         T::Item: Borrow<BufferImageCopy>;
 
-    ///
+    // TODO: This explanation needs improvement.
+    /// Performs a non-indexed drawing operation, drawing the given range
+    /// of vertices from the current vertex buffer.  It performs instanced
```
Comment (msiglreith) [inline src/hal/src/command/raw.rs]: (pedantic note: fetching attributes from the currently bound vertex buffers)

File: src/hal/src/command/raw.rs [hunk 12]
```diff
@@ -408,22 +441,38 @@ pub trait RawCommandBuffer<B: Backend>: Clone + Any + Send + Sync {
...
+    /// drawing, drawing `instances.len()`
+    /// times with an `instanceIndex` starting with the start of the range.
     fn draw(
         &mut self,
         vertices: Range<VertexCount>,
         instances: Range<InstanceCount>,
     );
 
+    /// Performs indexed drawing, drawing the given range of indices 
+    /// from the current vertex and index buffers.  `base_vertex` specifies
```
Comment (msiglreith) [inline src/hal/src/command/raw.rs]: single index buffer, zero or multiple vertex buffers
nit: whitespaces

File: src/hal/src/command/raw.rs [hunk 13]
```diff
@@ -441,19 +496,23 @@ pub trait RawCommandBuffer<B: Backend>: Clone + Any + Send + Sync {
...
+    /// Reset/clear the values in the given range of the query pool.
     fn reset_query_pool(&mut self, pool: &B::QueryPool, queries: Range<QueryId>);
 
-    ///
+    /// Requests a timestamp to be written.
     fn write_timestamp(&mut self, pso::PipelineStage, Query<B>);
 
-    ///
+    /// Modify constant data in a graphics pipeline.
+    /// Push constants are intended to modify data in a pipeline more
```
Comment (msiglreith) [inline src/hal/src/command/raw.rs]: data in the pipeline layout (AMD uses the term `user data`). It's basically a mem copy, the alternative would be updating the values inside a descriptor set. The better performance stems from the addtional indirection due to the descriptor sets setting inside the 'user data' and the shader has to dereference the descriptor set pointer

File: src/hal/src/command/render_pass.rs [hunk 14]
```diff
@@ -187,12 +195,15 @@ impl<'a, B: Backend, L: Level> Drop for RenderPassInlineEncoder<'a, B, L> {
     }
 }
 
-///
+/// An object that records commands into a command buffer where each command must
+/// be
```
Comment (msiglreith) [inline src/hal/src/command/render_pass.rs]: must be ..?

File: src/hal/src/format.rs [hunk 15]
```diff
@@ -96,11 +118,11 @@ impl Default for Swizzle {
 #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct Properties {
-    ///
+    /// A bitmask of the features supported when a device with linear tiling is requested.
```
Comment (msiglreith) [inline src/hal/src/format.rs]: tiling refers to images not devices

File: src/hal/src/image.rs [hunk 16]
```diff
@@ -529,67 +535,81 @@ impl From<RenderDesc> for DepthStencilDesc {
     }
 }
 
+/// Specifies options for how memory for an image is arranged.
+/// These are hints to the GPU driver and may or may not have actual
+/// performance effects, but describe constraints on how the data
+/// may be used that a program *must* obey.  They do not specify
```
Comment (msiglreith) [inline src/hal/src/image.rs]: nit: whitespaces

File: src/hal/src/image.rs [hunk 17]
```diff
@@ -529,67 +535,81 @@ impl From<RenderDesc> for DepthStencilDesc {
...
         /// Write-only state but can be combined with `COLOR_ATTACHMENT_READ`.
         const COLOR_ATTACHMENT_WRITE = 0x2;
-        ///
+        /// Read access to the buffer in a copy operation.
         const TRANSFER_READ = 0x4;
-        /// Write-only state of copy commands.
+        /// Write access to the buffer in a copy operation.
         const TRANSFER_WRITE = 0x8;
-        /// Read-only state for SRV access, or combine with `SHADER_WRITE` to have r/w access to UAV.
+        /// Read-only state for shader access, or combine with `SHADER_WRITE` to have r/w access to UAV.
```
Comment (msiglreith) [inline src/hal/src/image.rs]: hmm I actually chose the `SRV` term to differ it from `UAV` read access, which we currently can't handle with `SHADER_READ` only.

File: src/hal/src/lib.rs [hunk 18]
```diff
@@ -349,7 +352,7 @@ pub trait Backend: 'static + Sized + Eq + Clone + Hash + fmt::Debug + Any + Send
     type QueryPool:           fmt::Debug + Any + Send + Sync;
 }
 
-#[allow(missing_docs)]
+/// Marks that an error occured submitting  a
```
Comment (msiglreith) [inline src/hal/src/lib.rs]: submittting a  ..?

File: src/hal/src/pass.rs [hunk 19]
```diff
@@ -35,33 +35,37 @@ pub type AttachmentLayout = image::ImageLayout;
 #[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct AttachmentOps {
-    ///
+    /// Whether or not data from the load operation will be preserved after the subpass.
```
Comment (msiglreith) [inline src/hal/src/pass.rs]: That description is a bit confusing. It indicates how the data of the attachment will be load at first usage at the beginning of the subpass.

File: src/hal/src/pass.rs [hunk 20]
```diff
@@ -85,37 +89,42 @@ pub type AttachmentId = usize;
...
 pub type AttachmentRef = (AttachmentId, AttachmentLayout);
 
-///
+/// What other subpasses a particular subpass depends on.
 #[derive(Copy, Clone, Debug, Hash, PartialEq)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub enum SubpassRef {
-    ///
+    /// The subpass depends on something that was submitted to the
+    /// queue before the render pass began.
```
Comment (msiglreith) [inline src/hal/src/pass.rs]: before or after

File: src/hal/src/pass.rs [hunk 21]
```diff
@@ -85,37 +89,42 @@ pub type AttachmentId = usize;
...
     External,
-    ///
+    /// The subpass depends on another subpass with the given index.
```
Comment (msiglreith) [inline src/hal/src/pass.rs]: Probably worth noting that index must be smaller than or equal to the index of the current subpass.
Index refers here to the corresponding subpassdescription index in the renderpass creation

File: src/hal/src/pass.rs [hunk 22]
```diff
@@ -85,37 +89,42 @@ pub type AttachmentId = usize;
...
     Pass(usize),
 }
 
-/// Specifies dependencies between subpasses.
+/// Expresses a dependency between multiple subpasses.  This is used
```
Comment (msiglreith) [inline src/hal/src/pass.rs]: nit: whitespaces

File: src/hal/src/pass.rs [hunk 23]
```diff
@@ -85,37 +89,42 @@ pub type AttachmentId = usize;
...
 /// Description of a subpass for renderpass creation.
 pub struct SubpassDesc<'a> {
-    ///
+    /// Which attachments will be used as color buffers.
     pub colors: &'a [AttachmentRef],
-    ///
+    /// Which attachments will be used as depth/stencil buffers.
     pub depth_stencil: Option<&'a AttachmentRef>,
-    ///
+    /// Which attachments MAY be used by this subpass.
```
Comment (msiglreith) [inline src/hal/src/pass.rs]: may?

File: src/hal/src/pso/descriptor.rs [hunk 24]
```diff
@@ -1,4 +1,9 @@
 //! Descriptor sets and layouts.
+//! A descriptor is an object that describes the connection between a resource, such as
+//! an `Image` or `Buffer`, and a variable in a shader.  Descriptors are organized into
```
Comment (msiglreith) [inline src/hal/src/pso/descriptor.rs]: nit: whitespaces

File: src/hal/src/pso/descriptor.rs [hunk 25]
```diff
@@ -98,19 +103,24 @@ pub trait DescriptorPool<B: Backend>: Send + Sync + fmt::Debug {
         layouts.into_iter().map(|layout| self.allocate_set(layout.borrow())).collect()
     }
 
-    ///
+    /// Resets a descriptor pool, releasing all resources from all the descriptor sets
+    /// allocated from it and freeing the descriptor sets.  Invalidates all descriptor
```
Comment (msiglreith) [inline src/hal/src/pso/descriptor.rs]: nit: whitespaces

File: src/hal/src/pso/graphics.rs [hunk 26]
```diff
@@ -6,44 +6,63 @@ use super::input_assembler::{AttributeDesc, InputAssemblerDesc, VertexBufferDesc
...
     pub shaders: GraphicsShaderSet<'a, B>,
     /// Rasterizer setup
     pub rasterizer: Rasterizer,
     /// Vertex buffers (IA)
     pub vertex_buffers: Vec<VertexBufferDesc>,
     /// Vertex attributes (IA)
     pub attributes: Vec<AttributeDesc>,
-    ///
+    /// Input assembler attributes, describes how
+    /// vertices are assembled into triangles.
```
Comment (msiglreith) [inline src/hal/src/pso/graphics.rs]: primitives better than triangles imo

File: src/hal/src/pso/graphics.rs [hunk 27]
```diff
@@ -84,17 +104,22 @@ pub enum PolygonMode {
     Fill,
 }
 
-/// Which face, if any, to cull.
+/// Which faces of the polygon, if any, to cull.  Face culling
```
Comment (msiglreith) [inline src/hal/src/pso/graphics.rs]: nit: whitespaces

File: src/hal/src/pso/graphics.rs [hunk 28]
```diff
@@ -84,17 +104,22 @@ pub enum PolygonMode {
...
-    ///
+    /// Cull front faces
     Front,
-    ///
+    /// Cull back faces.
     Back,
 }
 
-/// The front face winding order of a set of vertices.
+/// The front face winding order of a set of vertices.  This is
```
Comment (msiglreith) [inline src/hal/src/pso/graphics.rs]: nit: whitespaces

File: src/hal/src/pso/graphics.rs [hunk 29]
```diff
@@ -104,15 +129,21 @@ pub enum FrontFace {
     CounterClockwise,
 }
 
+/// A depth bias allows changing the produced depth values 
+/// for fragments slightly but consistently.  This permits
```
Comment (msiglreith) [inline src/hal/src/pso/graphics.rs]: nit: whitespaces

File: src/hal/src/pso/graphics.rs [hunk 30]
```diff
@@ -126,11 +157,13 @@ pub struct Rasterizer {
...
-    ///
+    /// Whether or not to enable depth clamping; when enabled, instead of
+    /// fragments being omitted when they are outside the bounds of the z-plane,
+    /// they will be clamped to the min or max z value.
     pub depth_clamping: bool,
-    ///
+    /// What depth bias, if any, to use for the drawn primitives.
     pub depth_bias: Option<DepthBias>,
-    ///
+    /// DOC TODO: Not sure what this is intended for, seems unused.
```
Comment (msiglreith) [inline src/hal/src/pso/graphics.rs]: dx12 only feature, controls how triangles will be rasterized depending on their overlap with pixels
currently not implemented afaik

File: src/hal/src/pso/graphics.rs [hunk 31]
```diff
@@ -147,52 +180,38 @@ impl Rasterizer {
...
 }
 
-///
+/// A description of an equation for how to blend transparent, overlapping fragments.
 #[derive(Clone, Debug, Default, PartialEq)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub struct BlendDesc {
-    ///
+    /// DOC TODO: ???
+    /// Is this for this feature: https://msdn.microsoft.com/en-us/library/windows/desktop/bb205072(v=vs.85).aspx#Alpha_To_Coverage ?
```
Comment (msiglreith) [inline src/hal/src/pso/graphics.rs]: y

File: src/hal/src/pso/graphics.rs [hunk 32]
```diff
@@ -147,52 +180,38 @@ impl Rasterizer {
...
     pub alpha_coverage: bool,
-    ///
+    /// The logi operation to apply to the blending equation, if any.
```
Comment (msiglreith) [inline src/hal/src/pso/graphics.rs]: logic

File: src/hal/src/pso/input_assembler.rs [hunk 33]
```diff
@@ -48,25 +49,31 @@ pub struct AttributeDesc {
     pub element: Element<format::Format>,
 }
 
+/// Describes whether or not primitive restart is supported for
+/// an input assembler.  Primitive restart is a feature that
```
Comment (msiglreith) [inline src/hal/src/pso/input_assembler.rs]: nit: whitespaces

File: src/hal/src/pso/input_assembler.rs [hunk 34]
```diff
@@ -48,25 +49,31 @@ pub struct AttributeDesc {
...
+/// See <https://www.khronos.org/opengl/wiki/Vertex_Rendering#Primitive_Restart>
+/// for more detail.
 #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 pub enum PrimitiveRestart {
-    ///
+    /// No primitive restart.
     Disabled,
-    ///
+    /// Primitive restart using a 16-bit index value.
```
Comment (msiglreith) [inline src/hal/src/pso/input_assembler.rs]: using u16::max (0xFFFF) as primitive restart value
same below

File: src/hal/src/pso/mod.rs [hunk 35]
```diff
@@ -192,11 +195,20 @@ pub enum BasePipeline<'a, P: 'a> {
...
     Index(usize),
-    ///
+    /// No parent pipeline exists.
     None,
 }
 
 /// Specialization information for pipelines.
+/// 
+/// Specialization constants allow for easy configuration of 
+/// multiple similar pipelines.  For example, there may be a
```
Comment (msiglreith) [inline src/hal/src/pso/mod.rs]: nit: whitespaces

File: src/hal/src/pso/output_merger.rs [hunk 36]
```diff
@@ -258,6 +270,11 @@ pub struct StencilFace {
     pub op_pass: StencilOp,
 }
 
+/// Defines a stencil test.  Stencil testing is an operation
+/// performed after a fragment is created by a fragment shader;
```
Comment (msiglreith) [inline src/hal/src/pso/output_merger.rs]: Stencil test can also be performed early

File: src/hal/src/query.rs [hunk 37]
```diff
@@ -1,17 +1,21 @@
-//! Query operations
+//! Queries are commands that can be submitted to a command buffer to record statistics or
+//! other useful values as the command buffer is running.  They are often intended for profiling
```
Comment (msiglreith) [inline src/hal/src/query.rs]: nit: whitespaces

File: src/hal/src/query.rs [hunk 38]
```diff
@@ -28,11 +32,15 @@ bitflags!(
 
 /// Type of queries in a query pool.
 pub enum QueryType {
-    /// Occlusion query.
+    /// Occlusion query.  Count the number of drawn samples between
```
Comment (msiglreith) [inline src/hal/src/query.rs]: nit: whitespaces (also below)

File: src/hal/src/queue/mod.rs [hunk 39]
```diff
@@ -57,7 +56,11 @@ pub trait RawCommandQueue<B: Backend>: Any + Send + Sync {
         IC: IntoIterator,
         IC::Item: Borrow<B::CommandBuffer>;
 
+    /// Presents the result of the queue to the given swapchains, after waiting on all the
+    /// semaphores given in `wait_semaphores`.  A given swapchain must not appear in this
```
Comment (msiglreith) [inline src/hal/src/queue/mod.rs]: nit: whitespaces

File: src/hal/src/queue/mod.rs [hunk 40]
```diff
@@ -96,7 +100,9 @@ impl<B: Backend, C> CommandQueue<B, C> {
         }
     }
 
-    ///
+    /// Presents the result of the queue to the given swapchains, after waiting on all the
+    /// semaphores given in `wait_semaphores`.  A given swapchain must not appear in this
```
Comment (msiglreith) [inline src/hal/src/queue/mod.rs]: nit: whitespaces

File: src/hal/src/queue/submission.rs [hunk 41]
```diff
@@ -101,9 +101,10 @@ where
         }
     }
 
-    /// Promote a submission to a higher capability type.
+    /// Promote a submission to a higher (more general) capability type.  For example,
```
Comment (msiglreith) [inline src/hal/src/queue/submission.rs]: nit: whitespaces

File: src/hal/src/window.rs [hunk 42]
```diff
@@ -58,17 +58,20 @@ use std::any::Any;
 use std::borrow::{Borrow, BorrowMut};
 use std::ops::Range;
 
-///
+/// An extent describes the size of a rectangle, such as
+/// a window or texture.  It's not used for referring to a
```
Comment (msiglreith) [inline src/hal/src/window.rs]: nit: whitespaces

File: src/hal/src/window.rs [hunk 43]
```diff
@@ -96,11 +99,15 @@ pub struct SurfaceCapabilities {
...
+/// on the display.
 pub trait Surface<B: Backend>: Any + Send + Sync {
     /// Retrieve the surface image kind.
     fn kind(&self) -> image::Kind;
 
-    /// Check if the queue family supports presentation for this surface.
+    /// Check if the queue family supports presentation to this surface.
+    ///
+    /// For example, a queue family with only the `queue::QueueType::Compute` 
+    /// capability will not allow drawing commands to be issued.
```
Comment (msiglreith) [inline src/hal/src/window.rs]: The second part seems unrelated

**Comment by kvark** [comment]
> fatal: unable to access 'https://github.com/gfx-rs/gfx.git/': Failed to connect to github.com port 443: Timed out
Command exited with code 128

AppVeyor is not in the mood

**Comment by kvark** [comment]
Bors r+
